Password fetching from Vault, along w/ required CLI updates

john-osullivan opened the pull request:
This has all been tested with the private transaction loop, although y'all might want to run it again on your own machines to confirm for yourselves.

## Motivation
Rather than passing passwords in plaintext from the command line, we want to:
- Get the password's name and location in a Vault server/engine from the CLI
- Use Vault's AWS auth method to sign into the Vault server
- Retrieve the password and use it to unlock the specific account.

## Change Summary
- **main.go** - Modifies the `startNode` method to first check if we're using a password file, and if not, sends password fetching to our new module.  Adds various safety checks to determine what type of node we're trying to spin up.  Minor refactor of how maker/voter accounts are unlocked.
- **passwords.go** - Handles password fetching in all cases: 
  - Checks the provided arguments to determine whether we're using a Vault password (i.e. all required Vault args are present)
  - If we aren't, looks for a plaintext password arg and throws an error if not present
  - If we are, uses the [tx-executor AWS login functions](https://github.com/Eximchain/eximchain-transaction-executor/blob/e237250ced19a1ba860d215a7e79872d6fa913ae/main.go#L38) @Lsquared13 wrote to sign in and then fetch the password
- **flags.go** & **usage.go** - Adds the four new Vault CLI args and hooks their documentation into the built-in help functions.

## Gotchas
- **Assumption** - Safety checks assume that every running instance of this code will be either a maker, voter, or observer, and that some account type must always be passed in.  I was under the impression that was true and added a safety check to be certain, but thought I should point it out to confirm.
- **Degradation** - The built-in `unlock` argument supports unlocking multiple accounts, while this password fetching implementation only supports one password.  We decided fetching multiple passwords was unnecessary for our current problems, but something worth nothing.

john-osullivan committed:
Passwords module which either grabs password from CLI arg, or uses Vault creds from CLI args to fetch proper password from Vault. No behavior change if Vault args are not specified.

john-osullivan committed:
Added required CLI flags for fetching values from Vault.  Uses defaults where possible, documents our canonical values.

john-osullivan committed:
Modifies the startNode method to use new passwords module to fetch passwords

john-osullivan committed:
Eliminated a stub I forgot about

Lsquared13 commented on cmd/utils/flags.go:
```diff
@@ -389,6 +389,27 @@ var (
 		Usage: "Path of thr constellation private config",
 		Value: "",
 	}
+	// Vault flags
```
Mention the expectations around which of these flags are set together. They should be told that e.g. setting vaultaddr but not vaultpasswordpath will not work. Ideally put it in one of the "Usage" strings

Lsquared13 replied:
Also if we expect this to fail outside of EC2, mention that as well

Lsquared13 commented on cmd/utils/flags.go:
```diff
@@ -389,6 +389,27 @@ var (
+	// Vault flags
+	VaultAddrFlag = cli.StringFlag{
+		Name:  "vaultaddr",
+		Usage: "Web address to a Hashicorp Vault installation holding passwords",
```
This sounds awkward to me, and possibly a bit imprecise ("installation" sounds like "installed software" rather than "running server"

Maybe something more like:

"DNS or IP address where a Hashicorp Vault server holding passwords can be reached"

Lsquared13 commented on cmd/geth/main.go:
```diff
@@ -301,37 +317,32 @@ func startNode(ctx *cli.Context, stack *node.Node) {
-		}
+	usingVoterAcct := ctx.GlobalIsSet(utils.VoteAccountFlag.Name)
+	usingBlockMakerAcct := ctx.GlobalIsSet(utils.VoteBlockMakerAccountFlag.Name)
+	if len(accounts) == 0 && !usingVoterAcct && !usingBlockMakerAcct {
```
Should we also errror if `usingVoterAcct && usingBlockMakerAcct`? Seems like that would always be some sort of user mistake, since the vault password feature doesn't support multiple addresses in use concurrently.

Lsquared13 commented on cmd/geth/main.go:
```diff
@@ -301,37 +317,32 @@ func startNode(ctx *cli.Context, stack *node.Node) {
+	usingVoterAcct := ctx.GlobalIsSet(utils.VoteAccountFlag.Name)
+	usingBlockMakerAcct := ctx.GlobalIsSet(utils.VoteBlockMakerAccountFlag.Name)
+	if len(accounts) == 0 && !usingVoterAcct && !usingBlockMakerAcct {
+		utils.Fatalf("Was not provided an `unlock`, `voteaccount`, or `blockmakeraccount` flag, cannot launch.")
```
Is there any chance this causes a node ignoring our vault password feature to crash, which didn't crash before?

john-osullivan replied:
Yeah, this is what I meant about my assumption that all nodes will be either observers, voters, or makers, respectively.  If a node isn't unlocking any accounts on startup, I'm explicitly throwing an error.

Is that not valid behavior w/ how this will be run?

Lsquared13 commented on cmd/geth/passwords.go:
```diff
@@ -0,0 +1,158 @@
+		// Authenticate to Vault via the AWS method
+		vaultConfig := vaultAPI.DefaultConfig()
+		vaultConfig.Address = ctx.GlobalString(utils.VaultAddrFlag.Name)
+		vaultClient, err := vaultAPI.NewClient(vaultConfig)
```
Check this for errors before overwriting `err` with the aws login

Lsquared13 commented on cmd/geth/passwords.go:
```diff
@@ -0,0 +1,158 @@
+		keyname := ctx.GlobalString(utils.VaultPasswordNameFlag.Name)
+		password, present := secret.Data[keyname]
+		if !present {
+			utils.Fatalf("fetchPasswordFromVault found a secret at specified path (%v), but secret did not contain specified key name (%v). Secret was : %v", fullSecretPath, keyname, secret.Data)
```
Let's not log the secret data?

john-osullivan replied:
Just like they say in the Navy:

> Loose log statements sink ships!

I think that's how it went, right?  Thanks for catching that, holdover from debugging.

Lsquared13 commented on cmd/geth/passwords.go:
```diff
@@ -0,0 +1,158 @@
+		if len(setPassFlags) == 1 {
+			return false
+		}
+		utils.Fatalf("Too many (%v) password flags have been set.  Only one of the following should be supplied: %v", len(setPassFlags), setPassFlags)
```
Can we avoid doing this in the case where none of the vault flags are set? Ideally we only lose multiple account capabilities if people try to use the vault password feature.

john-osullivan replied:
So this logic is actually outdated now, as the `fetchPassword` and `usingVaultPassword` functions will only get called if there's no password file provided, which is a key part of how multiple account unlocking works.  If plaintext-password flags are provided for both voting &  making, that seems like another instance of the error above where somebody provides both a voting address and a making address.

Updating logic to remove references to password file.

Lsquared13 commented on cmd/geth/passwords.go:
```diff
@@ -0,0 +1,158 @@
+}
+
+func loginAws(v *vaultAPI.Client) (string, error) {
+	loginData, err := awsauth.GenerateLoginData("", "", "", "")
```
What are those arguments? Can we put their names in block quotes?

e.g.
```
awsauth.GenerateLoginData(/*arg1=*/"", /*arg2=*/"", /*arg3=*/"", /*arg4=*/"")
```

Lsquared13 commented on cmd/geth/passwords.go:
```diff
@@ -0,0 +1,158 @@
+	if secret == nil {
+		return "", fmt.Errorf("empty response from credential provider")
+	}
+	if secret.Auth == nil {
```
Maybe check secret.Auth.ClientToken too?

Lsquared13 commented:
Regarding the multiple account thing and validation, my ideal implementation is "If none of our flags are mentioned, don't break anything that was working before. If any of our flags are mentioned, strictly validate that they adhere to all our assumptions and fail fast if they don't."

john-osullivan committed:
Updated flag usage text, made debug log stop leaking info, added more safety checks, removed logic suggesting password module would still be called if password file flag is provided

john-osullivan commented:
@Lsquared13 most recent commit addresses most of your comments.  The built-in unlocking multiple accounts functionality explicitly depends on the `passwordfile` flag being provided, as it expects each password to be a line.  If that flag is set, none of my `password` module checks which look into Vault will ever be called.

Right now, people are able to specify a password file for unlocking their voter/maker accounts.  This is the order that I look for passwords, and whichever one comes up first, we use for whichever accounts (voter, maker, observer) were requested:

1. Put that account password in a file and specify that file
2. Specify that password as a plaintext arg
3. Include Vault arguments to get the password from there

Should we throw an error if people try to use more than one of these strategies?

~~My check expects that nodes will always be started with an account to unlock.  Is that how things actually work?~~  **Edit:** Talked it over with Juan, sounds like that's not the case -- people *are* allowed to spin up their own node without unlocking any accounts, so I just pushed up another commit (below this comment) removing that check.

john-osullivan committed:
Removed fatal check that ensured some account was provided after confirming that a node can be run without unlocking any initial accounts

john-osullivan commented:
Hey, just following up here & bumping you both (@Lsquared13 @eximchain137 ) as it's the more serious PR and I'd love to finish this sprint with my first blocker done.  All the comments have been addressed at this point, the only thing left is making sure that the password handling is behaving as intended. Louis, your angle makes a lot of sense:

> "If none of our flags are mentioned, don't break anything that was working before. If any of our flags are mentioned, strictly validate that they adhere to all our assumptions and fail fast if they don't."

With that in mind, here are all the password fetching mechanisms in this version of Quorum, along with my assumptions about them and some italicized wrinkles to consider:
- **`--passwordfile`**: This is the only built-in password flag.  It was originally meant to be used with the `--unlock` arg, which lets you specify a comma-separated list of accounts to unlock (that's how the multiple account unlock functionality works).  However, we also use `--unlock` with our new Vault password, which does not support multiple accounts.
  - *We also use `--unlock` with our new Vault password, which does not support multiple accounts.  If both the Vault args & `--passwordfile` are present, should that be an error case?* 
  - *Should this arg let you unlock the `--voteaccount` or `--blockmakeraccount`?  That's technically possible with my implementation, but I could add a safety check preventing it.*
- **`--votepassword`/`--blockmakerpassword`**: These are our plaintext password flags.  Only one should be provided -- if both are, we throw an error.  If one is provided, we must also get the symmetric `account` flag or we throw an error.
  - *Should people be able to use both the `unlock`/`passwordfile` pair at the same time as a `voteaccount`/`votepassword` pair?  Or should that case cause an error?*
  - *If either of these pairs are present and we also detect a supplied Vault arg, should we throw a safety error?*
- **`--vaultaddr, prefix, passwordname, passwordpath`**: New args I'm adding; `prefix` and `passwordname` have default values.  This only works if all four are specified; if either the `addr` or `path` are supplied and the other is missing, we throw an error.
  - *If we find three of these specified, should we throw an error even if one of the other schemas is properly specified?*

Lsquared13 commented:
> We also use `--unlock` with our new Vault password, which does not support multiple accounts. If both the Vault args & `--passwordfile` are present, should that be an error case?

Yes. If the vault args are specified we should be aggressive about validating that they are using this in the expected manner

> Should this arg let you unlock the `--voteaccount` or `--blockmakeraccount?` That's technically possible with my implementation, but I could add a safety check preventing it

I'm not actually sure what vanilla quorum does but can we try to do what vanilla quorum does, since vanilla quorum also has those arguments along with the `-unlock` argument?

> Should people be able to use both the `unlock`/`passwordfile` pair at the same time as a `voteaccount`/`votepassword` pair? Or should that case cause an error?

Same as above. Since this doesn't involve our vault arguments at all can we do what vanilla quorum does? You'll need to verify what it does.

> If either of these pairs are present and we also detect a supplied Vault arg, should we throw a safety error?

Yes, in keeping with our "aggressive validation if you specify any of our new options" philosophy.

> If we find three of these specified, should we throw an error even if one of the other schemas is properly specified?

I'm not 100% sure what you're asking here, but the core principle is "If any of our arguments are actually mentioned, assert that all of them are set after accounting for defaults"

- If any flag is set, assert the others are set after accounting for defaults. We do not care how they are filled. 
- Checking if the flags with defaults have been changed from their defaults is an acceptable shortcut, we don't really need to catch the case where someone specifies a default for no reason.

john-osullivan committed:
Modified validation behavior such that any Vault args mean there should be no other password args.  Restored baseline to be identical with pre-Vault behavior.

john-osullivan committed:
Removed two fxns that are no longer required w new behavior

Lsquared13 commented on cmd/geth/main.go:
```diff
@@ -275,13 +279,35 @@ func startNode(ctx *cli.Context, stack *node.Node) {
+			if usingVault {
+				if usedVault {
+					utils.Fatalf("Vault passwords only support unlocking one account, it looks like you are trying to unlock %v.", len(accounts))
+				} else {
```
nit: get rid of the else block, just put `usedVault = true` under the if block. We hit that statement in every case we haven't already FATALed

Lsquared13 commented on cmd/geth/passwords.go:
```diff
@@ -0,0 +1,154 @@
+	// Login data args are left empty so that Vault's AWSAuth implementation
+	// knows to let AWS's EnvProvider handle it.  The EnvProvider searches the
+	// environment for these values: https://github.com/aws/aws-sdk-go/blob/master/aws/credentials/env_provider.go
+	loginData, err := awsauth.GenerateLoginData( /*accessKey=*/ "" /*secretKey=*/, "" /*sessionToken=*/, "" /*headerValue=*/, "")
```
nit: weird comma groupings

Change to:

```
loginData, err := awsauth.GenerateLoginData( /*accessKey=*/ "", /*secretKey=*/ "", /*sessionToken=*/ "", /*headerValue=*/ "")
```

john-osullivan replied:
Agreed that these groupings are weird, but that's what the go linter keeps setting them to when I save.  I'd rather not disable the linter, so for the sake of legibility, I'm throwing each arg-quote pair on its own line.

Lsquared13 commented on cmd/geth/passwords.go:
```diff
@@ -0,0 +1,154 @@
+	if secret.Auth == nil {
+		return "", fmt.Errorf("Secret contains no auth data.")
+	}
+	if secret.Auth.ClientToken == "" {
```
Do we also need to check for nil?

john-osullivan replied:
We actually can't check for nil, as that doesn't satisfy the type requirement of `Auth`.  I added this `""` check because a nil check wouldn't compile.

Lsquared13 reviewed (changes requested):
A few nits

john-osullivan committed:
Removed unnecessary else clause, cleaned up GenerateLoginData arglist

eximchain137 commented on cmd/geth/main.go:
```diff
@@ -301,32 +326,58 @@ func startNode(ctx *cli.Context, stack *node.Node) {
+		} else if usingMakerPassword {
 			passwd = append(passwd, ctx.GlobalString(utils.VoteBlockMakerAccountPasswordFlag.Name))
+		} else {
+			utils.Fatalf("Tried to unlock a voter account, but no password was provided via --blockmakerpassword or Vault password.")
```
Tried to unlock a *blockmaker* account, but no password was provided via --blockmakerpassword or Vault password.

john-osullivan replied:
Most recent commit fixes this

eximchain137 commented on cmd/geth/passwords.go:
```diff
@@ -0,0 +1,159 @@
+func fetchPasswordFromVault(ctx *cli.Context) (string, error) {
+	if usingVaultPassword(ctx) {
+		// Authenticate to Vault via the AWS method
+		vaultConfig := vaultAPI.DefaultConfig()
```
defaults to 60s time out and 3 retries for client make sure this is good cross region @Lsquared13

Lsquared13 replied:
Maybe do a cross-region test putting the nodes in one region and the primary-region in a region halfway around the world (e.g. us-east-1 and ap-southeast-2)?

john-osullivan replied:
Cool, I'll run a private tx test w/ those settings now

john-osullivan committed:
Added some explanatory comments, made sure Vault client cleared token after use

john-osullivan commented:
Okay, this most recent commit fixes an error message and adds a few more explanatory comments.  Still running my cross-world Vault connection test, I'll drop another comment once that's complete.  Otherwise, though, looks like this PR is good to go!

Lsquared13 reviewed (approved):
Pending success on the cross region test

john-osullivan commented:
Finally finished verifying this cross-region test -- ready to go!
